repentogon/Patches/ASMPatches/ASMStatusEffects.cpp

```diff
@@ -0,0 +1,169 @@
+	{"8bcee8????????84c074??8b86????????85c074??90",
+	"AddKnockback", 0, ASMPatch::Registers::ECX, ASMPatch::Registers::ESI, ASMPatch::SavedRegisters::Registers::ESI, 0x22},
+	{"e8????????84c074??8b8e????????85c974??0f1f44??008bf18b8e????????85c975??85d20f84????????8b86????????83e00183c8000f85????????3986????????0f8f????????8b8e????????85d27e??8d04??b92c0100003bc10f4cc8eb??f7da3bca0f4cca898e????????8d4c24??0f2805????????6a010f114424??6a000f2805",
+	"AddMagnetized", 0, ASMPatch::Registers::ECX, ASMPatch::Registers::ESI, ASMPatch::SavedRegisters::Registers::ESI, 0x24},
+	{"e8????????84c074??8b8e????????85c974??0f1f44??008bf18b8e????????85c975??85ff",
+	"AddMidasFreeze", 0, ASMPatch::Registers::ECX, ASMPatch::Registers::ESI, ASMPatch::SavedRegisters::Registers::ESI, 0x24},
+
+	// this crashed on me once, but now i can't replicate it
+	{"e8????????84c074??8b86????????85c074??0f1f80000000008bf08b86????????85c075??85d20f84????????8bbe????????8bc78b9e????????83e00183c8000f85????????3986????????0f8f????????8b8e????????85d27e??8d04??b9780000003bc10f4cc8eb??f7da3bca0f4cca8bc7898e????????83e040",
+	"AddPoison", 0, ASMPatch::Registers::ECX, ASMPatch::Registers::ESI, ASMPatch::SavedRegisters::Registers::ESI, 0x26},
+
+	{"8bf7e8????????84c074??8b8f????????85c974??8bf18b8e????????85c975??8b8e",
+	"AddShrink", 0, ASMPatch::Registers::ECX, ASMPatch::Registers::ESI, ASMPatch::SavedRegisters::Registers::ESI, 0x21},
+	{"e8????????84c074??8b8e????????85c974??0f1f00",
+	"AddSlowing", 0, ASMPatch::Registers::ECX, ASMPatch::Registers::ESI, ASMPatch::SavedRegisters::Registers::ESI, 0x22},
+	{"e8????????84c074??8b8e????????85c974??8bf1",
+	"AddWeakness", 0, ASMPatch::Registers::ECX, ASMPatch::Registers::ESI, ASMPatch::SavedRegisters::Registers::ESI, 0x1f},
+};
+
+/* /////////////////////
+// shared GetStatusEffectTarget trampoline
+*/ /////////////////////
+
+Entity* __stdcall GetStatusEffectTargetTrampoline(Entity* target) {
+	return target->GetStatusEffectTarget();
+}
+
+/* /////////////////////
+// Generic inline patch
+*/ /////////////////////
+
+void ASMPatchInlinedGetStatusEffectTarget(void* addr, ASMPatch::Registers entityReg, ASMPatch::Registers targetReg, ASMPatch::SavedRegisters::Registers savedReg, unsigned int jumpOffset) {
+	ASMPatch::SavedRegisters savedRegisters(ASMPatch::SavedRegisters::Registers::GP_REGISTERS + ASMPatch::SavedRegisters::XMM1 - savedReg, true);
+	ASMPatch patch;
+	patch.PreserveRegisters(savedRegisters)
+		.Push(entityReg)
+		.AddInternalCall(GetStatusEffectTargetTrampoline)
+		// poor man's mov reg,reg
+		.Push(ASMPatch::Registers::EAX)
+		.Pop(targetReg)
+		.RestoreRegisters(savedRegisters)
+		.AddRelativeJump((char*)addr + jumpOffset);
+	sASMPatcher.PatchAt(addr, &patch);
+}
+
+/* /////////////////////
+// Thank you, God bless you, and God bless the United States of America.
+*/ /////////////////////
+
+bool avoidIceCrash = false;
```

Comment:
Please add a comment in the code explaining why we need this hack, because right now I'm terrified of what releasing this feature will do

Reply:
One specific call to AddKnockback when colliding with iced enemies was somehow leading to ECX containing 0x0 when calling GetType, which of course doesn't work well. I do agree in retrospect that this solution is a bit maniacal. I will try to work around it a different way.

Reply:
A more proper fix has been found and implemented.